### ISSUE TITLE
Small enhancements please

### ISSUE THREAD

**Issue by ned14** [meqif/rust-utp#3]
Some really small stuff:

* I can't convert an existing UtpSocket into a UtpStream (no suitable constructor).

* If I bind to 0.0.0.0:0, the underlying UdpSocket will choose a random port. I have no way of retrieving that random port allocation from UtpSocket nor UtpListener currently.

* UtpListener::accept() doesn't return a (UtpSocket, SocketAddr) like TcpListener does.

Thanks,
Niall

**Comment by meqif** [comment]
> I can't convert an existing UtpSocket into a UtpStream (no suitable constructor).

Actually, you can in the master branch. Just use `into()`.

> If I bind to 0.0.0.0:0, the underlying UdpSocket will choose a random port. I have no way of retrieving that random port allocation from UtpSocket nor UtpListener currently.

Fair enough. Would `fn local_addr(&self) -> Result<SocketAddr>` (like in UdpSocket) do?

> UtpListener::accept() doesn't return a (UtpSocket, SocketAddr) like TcpListener does.

I knew I forgot something when I implemented `accept`.

Alright, I'll have a new minor release ready with those issues fixed probably today/tonight. Thanks for bringing up those issues.

**Comment by meqif** [comment]
I got things done sooner than I expected. Check out 0.5.0.

**Comment by ned14** [comment]
Oh wow. Now that is service!

You've just unblocked me here as I am getting crust to create two utp connections per connection as crust needs to concurrently read and write, and UtpSocket can't do concurrency nor does it allow cloning like Tcp. However a much nicer alternative would be either some way of interrupting blocked reads to do a write, and then resume the read, or else run a mutex around the utp management code as we discussed before. I appreciate both are probably quite a lot of work.

I just wish Rust could multiplex socket waits.

**Comment by meqif** [comment]
Glad to be of service.

Yes, the lack of non-blocking reads on sockets has been bothering me for a while. That would avoid a lot of pain.

I could rely on [mio](https://github.com/carllerche/mio) for that, but that would kill support for Windows, which I'd like to avoid.
I tried the mutex approach some months ago and got a working prototype, but the code was too messy for my taste. It might work with careful design. 

Cloning probably wouldn't work correctly because the internal state would get out of sync very quickly. In fact, there used to be `Clone` implementations for both `UtpStream` and `UtpSocket`, but they had to be removed partly because of that.

Maybe a `try_recv` method similar to [`mpsc::Receiver`'s](http://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.try_recv) would be useful? It would either return buffered data (received while sending packets) or `WouldBlock` (or a similar error). It wouldn't help with a socket stuck on receiving, but it could be used between sends to drain the socket/stream's incoming buffer.

**Comment by ned14** [comment]
What we have in crust is a dedicated thread which reads, and a dedicated thread which writes. Those are internal, so they could be the same thread. One hacky way one could work around the blocking UtpSocket read is to internally send a null length udp datagram to the listening udp socket before each write. That would unblock the read, let us do the write, and then one could resume the blocking read. Food for thought.

P.S. I absolutely agree that Rust needs some very serious improvement in its sockets implementation. I personally find ASIO in C++ a bit clunky and old fashioned, however dedicating kernel threads to block on i/o hasn't been necessary on any system except on Linux for well over a decade now. Anyway, we play we the hand we're dealt.

**Comment by ned14** [comment]
BTW UtpListener is missing a local_addr() :)

**Comment by meqif** [comment]
Oops. Now it isn't. Check out 0.5.1. :wink:

**Comment by ned14** [comment]
Thanks!

I'm up to the point in crust where I can go no further without rust-utp letting me do sends during concurrent reads. I have no problem doing the work on rust-utp myself, but just to check which you would prefer:

1. Would you prefer a solution whereby a new UtpReadStream and a UtpWriteStream would be added. The UtpReadStream takes ownership of a UtpSocket, whilst a UtpWriteStream takes a reference to its corresponding UtpReadStream. The idea is that two threads can each take their UtpReadStream and UtpWriteStream instances.

2. The earlier isn't Rust idomatic, so you might prefer that UtpStream becomes cloneable, then as many instances of UtpStream can exist as desired. UtpSocket remains non thread safe.

I was thinking of implementing concurrency by simply having UdpStream spawn a thread peeking at the internal UdpSocket. When the UdpSocket becomes ready with new data, it locks the stream and does the real UdpSocket read knowing it will never block, and then wakes any readers waiting on reading the UdpStream.

This design is very non-invasive to your UdpSocket, which is why I chose it. But if you'd prefer a different and/or better design, please do speak up before I fork rust-utp on Monday. I believe MaidSafe are quite flexible as to helping you out with whatever design you prefer.

Niall

**Comment by meqif** [comment]
I see. I'd prefer to keep the API as close to `TcpStream` as possible, as you've probably noticed, so the second way seems better to me. I need to think some more about this, though.

I tried playing a bit with a socket wrapped in an Arc and a Mutex (`Arc::new(Mutex::new(socket))`) and a listening thread, and it seemed to work reasonably well without changes in the library, but that was a toy.

To keep things clean, would you mind opening a separate issue with that comment?

By the way, 0.6.0 is out. `UtpSocket::connect` is now more similar to its `TcpStream` counterpart — it's an associated function instead of a method, which avoids the `bind` call before `connect`. That's a breaking change, although a small one.
